File: quickshell/aeris-dashboard/services/weather.py
#!/usr/bin/python3
"""Open-Meteo current model conditions, with bounded, location-keyed disk cache."""
import argparse
import json
import math
import os
from pathlib import Path
import tempfile
import time
from urllib.parse import urlencode
from urllib.request import Request, urlopen

MAX_AGE = 6 * 3600
REFRESH = 600
ERRORS = (OSError, ValueError, KeyError, TypeError)


def location():
    path = Path(os.environ.get("AERIS_WEATHER_CONFIG", str(Path(os.environ.get(
        "XDG_CONFIG_HOME", str(Path.home() / ".config"))) / "aeris-dashboard/weather.json")))
    data = json.loads(path.read_text())
    for field, limit in (("latitude", 90), ("longitude", 180)):
        value = data[field]
        if type(value) not in (int, float) or not math.isfinite(value) or abs(value) > limit:
            raise ValueError("Invalid weather coordinates")
    if not isinstance(data.get("name"), str) or not 1 <= len(data["name"]) <= 80:
        raise ValueError("Invalid weather location name")
    return {key: data[key] for key in ("name", "latitude", "longitude")}


def condition(code, day):
    if code in (0, 1):
        return ("Clear sky" if code == 0 else "Mainly clear", "clear" if day else "night",
                "white-balance-sunny" if day else "moon-waning-crescent")
    if code == 2:
        return ("Partly cloudy", "partly-cloudy", "weather-partly-cloudy" if day else "weather-night-partly-cloudy")
    if code == 3:
        return ("Overcast", "cloudy", "weather-cloudy")
    if code in (45, 48):
        return ("Fog", "fog", "weather-fog")
    if code in (51, 53, 55, 56, 57):
        return ("Freezing drizzle" if code in (56, 57) else "Drizzle", "rain", "weather-rainy")
    if code in (61, 63, 65, 66, 67, 80, 81, 82):
        label = "Freezing rain" if code in (66, 67) else "Showers" if code >= 80 else "Rain"
        return (label, "rain", "weather-rainy")
    if code in (71, 73, 75, 77, 85, 86):
        return ("Snow showers" if code >= 85 else "Snow", "snow", "weather-snowy")
    if code in (95, 96, 99):
        return ("Thunder + hail" if code in (96, 99) else "Thunderstorm", "storm", "weather-lightning-rainy")
    return ("Unknown conditions", "unknown", "weather-cloudy")


def normalize(data, now):
    current = data["current"]
    temperature = current["temperature_2m"]
    code, day, measured = current["weather_code"], current["is_day"], current["time"]
    if type(temperature) not in (int, float) or not math.isfinite(temperature) or not -100 <= temperature <= 70:
        raise ValueError("Invalid weather temperature")
    if type(code) is not int or type(day) is not int or day not in (0, 1):
        raise ValueError("Invalid weather condition")
    if type(measured) not in (int, float) or not math.isfinite(measured) or not -300 <= now - measured < MAX_AGE:
        raise ValueError("Weather provider returned an expired timestamp")
    if data["current_units"]["temperature_2m"] != "°C":
        raise ValueError("Unexpected weather temperature unit")
    label, pattern, icon = condition(code, day)
    return {"temperature": temperature, "code": code, "isDay": bool(day),
            "description": label, "condition": pattern, "icon": icon,
            "observedAt": measured, "fetchedAt": now}


def fetch(place, now):
    query = urlencode({"latitude": place["latitude"], "longitude": place["longitude"],
        "current": "temperature_2m,weather_code,is_day", "temperature_unit": "celsius",
        "timeformat": "unixtime", "timezone": "auto", "forecast_days": 1})
    req = Request("https://api.open-meteo.com/v1/forecast?" + query,
                  headers={"User-Agent": "Aeris-Weather/1.0"})
    with urlopen(req, timeout=12) as response:
        payload = response.read(131073)
    if len(payload) > 131072:
        raise ValueError("Weather response too large")
    return normalize(json.loads(payload), now)


def cache_path():
    return Path(os.environ.get("AERIS_WEATHER_CACHE", str(Path(os.environ.get(
        "XDG_CACHE_HOME", str(Path.home() / ".cache"))) / "aeris-dashboard/weather.json")))


def read_cache(place, now):
    try:
        cached = json.loads(cache_path().read_text())
        if cached["location"] != place:
            return None
        value = cached["weather"]
        # Revalidate stored numbers/mapping, rather than trusting a partial cache.
        result = normalize({"current": {"temperature_2m": value["temperature"],
            "weather_code": value["code"], "is_day": int(value["isDay"]),
            "time": value["observedAt"]}, "current_units": {"temperature_2m": "°C"}}, now)
        age = now - value["fetchedAt"]
        if not 0 <= age < MAX_AGE:
            return None
        result["fetchedAt"] = value["fetchedAt"]
        return result
    except ERRORS:
        return None


def save_cache(place, weather):
    path = cache_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, staging = tempfile.mkstemp(dir=path.parent, prefix=".weather-")
    try:
        with os.fdopen(fd, "w") as output:
            json.dump({"location": place, "weather": weather}, output)
        os.replace(staging, path)
    finally:
        if os.path.exists(staging):
            os.unlink(staging)


def collect(force=False):
    now = time.time()
    try:
        place = location()
    except ERRORS:
        return {"ok": False, "available": False, "description": "Set location",
                "error": "Configure ~/.config/aeris-dashboard/weather.json"}
    cached = read_cache(place, now)
    if not force and cached and now - cached["fetchedAt"] < REFRESH:
        return {**cached, "ok": True, "available": True, "location": place["name"]}
    try:
        weather = fetch(place, now)
    except ERRORS:
        return {**(cached or {}), "ok": False, "available": cached is not None,
                "location": place["name"], "error": "Weather update unavailable",
                **({} if cached else {"description": "Unavailable"})}
    try:
        save_cache(place, weather)
    except OSError:
        pass  # A read-only cache must not discard a successful network response.
    return {**weather, "ok": True, "available": True, "location": place["name"]}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--refresh", action="store_true")
    print(json.dumps(collect(parser.parse_args().refresh), allow_nan=False))

File: quickshell/aeris-backend/src/weather.rs
use crate::common::{self, Result, err};
use serde_json::{Value, json};
use std::{env, path::PathBuf, time::Duration};
const MAX_AGE: f64 = 21600.0;
pub fn paths() -> (PathBuf, PathBuf) {
    (
        env::var_os("AERIS_WEATHER_CONFIG")
            .map(PathBuf::from)
            .unwrap_or_else(|| {
                common::xdg("XDG_CONFIG_HOME", ".config").join("aeris-dashboard/weather.json")
            }),
        env::var_os("AERIS_WEATHER_CACHE")
            .map(PathBuf::from)
            .unwrap_or_else(|| {
                common::xdg("XDG_CACHE_HOME", ".cache").join("aeris-dashboard/weather.json")
            }),
    )
}
pub fn location(data: &Value) -> Result<Value> {
    for (field, limit) in [("latitude", 90.0), ("longitude", 180.0)] {
        if !data[field]
            .as_f64()
            .is_some_and(|n| n.is_finite() && n.abs() <= limit)
        {
            return Err("Invalid weather coordinates".into());
        }
    }
    let name = data["name"]
        .as_str()
        .filter(|s| (1..=80).contains(&s.chars().count()))
        .ok_or("Invalid weather location name")?;
    Ok(json!({"name":name,"latitude":data["latitude"],"longitude":data["longitude"]}))
}
pub fn condition(code: i64, day: bool) -> (&'static str, &'static str, &'static str) {
    match code {
        0 | 1 => (
            if code == 0 {
                "Clear sky"
            } else {
                "Mainly clear"
            },
            if day { "clear" } else { "night" },
            if day {
                "white-balance-sunny"
            } else {
                "moon-waning-crescent"
            },
        ),
        2 => (
            "Partly cloudy",
            "partly-cloudy",
            if day {
                "weather-partly-cloudy"
            } else {
                "weather-night-partly-cloudy"
            },
        ),
        3 => ("Overcast", "cloudy", "weather-cloudy"),
        45 | 48 => ("Fog", "fog", "weather-fog"),
        51 | 53 | 55 | 56 | 57 => (
            if code >= 56 {
                "Freezing drizzle"
            } else {
                "Drizzle"
            },
            "rain",
            "weather-rainy",
        ),
        61 | 63 | 65 | 66 | 67 | 80 | 81 | 82 => (
            if code == 66 || code == 67 {
                "Freezing rain"
            } else if code >= 80 {
                "Showers"
            } else {
                "Rain"
            },
            "rain",
            "weather-rainy",
        ),
        71 | 73 | 75 | 77 | 85 | 86 => (
            if code >= 85 { "Snow showers" } else { "Snow" },
            "snow",
            "weather-snowy",
        ),
        95 | 96 | 99 => (
            if code == 95 {
                "Thunderstorm"
            } else {
                "Thunder + hail"
            },
            "storm",
            "weather-lightning-rainy",
        ),
        _ => ("Unknown conditions", "unknown", "weather-cloudy"),
    }
}
pub fn normalize(data: &Value, now: f64) -> Result<Value> {
    let current = &data["current"];
    let temp = current["temperature_2m"]
        .as_f64()
        .filter(|n| n.is_finite() && (-100.0..=70.0).contains(n))
        .ok_or("Invalid weather temperature")?;
    let code = current["weather_code"]
        .as_i64()
        .ok_or("Invalid weather condition")?;
    let day = current["is_day"]
        .as_i64()
        .filter(|n| *n == 0 || *n == 1)
        .ok_or("Invalid weather day flag")?;
    let measured = current["time"]
        .as_f64()
        .filter(|n| n.is_finite() && (-300.0..MAX_AGE).contains(&(now - n)))
        .ok_or("Expired weather timestamp")?;
    if data["current_units"]["temperature_2m"] != "°C" {
        return Err("Unexpected temperature unit".into());
    }
    let (label, pattern, icon) = condition(code, day == 1);
    Ok(
        json!({"temperature":temp,"code":code,"isDay":day==1,"description":label,"condition":pattern,"icon":icon,"observedAt":measured,"fetchedAt":now}),
    )
}
pub fn cached(data: &Value, place: &Value, now: f64) -> Option<Value> {
    if data["location"] != *place {
        return None;
    }
    let old = &data["weather"];
    let mut value = normalize(&json!({"current":{"temperature_2m":old["temperature"],"weather_code":old["code"],
        "is_day":if old["isDay"].as_bool()? {1} else {0},"time":old["observedAt"]},"current_units":{"temperature_2m":"°C"}}),now).ok()?;
    let fetched = old["fetchedAt"].as_f64()?;
    if !(0.0..MAX_AGE).contains(&(now - fetched)) {
        return None;
    }
    value["fetchedAt"] = old["fetchedAt"].clone();
    Some(value)
}
pub fn collect_with(
    config: &std::path::Path,
    cache: &std::path::Path,
    now: f64,
    force: bool,
    fetch: impl FnOnce(&Value, f64) -> Result<Value>,
) -> Value {
    let Ok(place) = common::read_json(config, 65536).and_then(|v| location(&v)) else {
        return json!({"ok":false,"available":false,"description":"Set location","error":"Configure ~/.config/aeris-dashboard/weather.json"});
    };
    let old = common::read_json(cache, 131072)
        .ok()
        .and_then(|v| cached(&v, &place, now));
    if !force
        && let Some(mut value) = old.clone()
        && now - value["fetchedAt"].as_f64().unwrap_or(0.0) < 600.0
    {
        value["ok"] = json!(true);
        value["available"] = json!(true);
        value["location"] = place["name"].clone();
        return value;
    }
    match fetch(&place, now) {
        Ok(mut value) => {
            let _ = common::atomic_json(cache, &json!({"location":place,"weather":value}), false);
            value["ok"] = json!(true);
            value["available"] = json!(true);
            value["location"] = place["name"].clone();
            value
        }
        Err(_) => {
            let available = old.is_some();
            let mut value = old.unwrap_or_else(|| json!({"description":"Unavailable"}));
            value["ok"] = json!(false);
            value["available"] = json!(available);
            value["location"] = place["name"].clone();
            value["error"] = json!("Weather update unavailable");
            value
        }
    }
}
pub fn collect(force: bool) -> Value {
    let (config, cache) = paths();
    collect_with(&config, &cache, common::now(), force, |place, now| {
        let url = format!(
            "https://api.open-meteo.com/v1/forecast?latitude={}&longitude={}&current=temperature_2m%2Cweather_code%2Cis_day&temperature_unit=celsius&timeformat=unixtime&timezone=auto&forecast_days=1",
            place["latitude"], place["longitude"]
        );
        let bytes = common::get(&url, Duration::from_secs(12), 131072, "Aeris-Weather/1.0")?;
        normalize(&serde_json::from_slice(&bytes).map_err(err)?, now)
    })
}
